src/gui_fb.c

```diff
@@ -0,0 +1,874 @@
+    uint32_t r, g, b, color;
+} Drawable, Texture;
+
+static struct termios orig;
+static int frame_fd = -1;           /* framebuffer file handle */
+static int mouse_fd = -1;
+static int keybd_fd = -1;;
+static int posx, posy;              /* cursor position */
+static Drawable fb;                 /* hardware framebuffer */
+static Drawable bb;                 /* back buffer */
+static Rect clip;
+static pixel_t draw_color = -1;      /* default color */
+
+void host_gui_deinit(void);
+static int open_framebuffer(void);
+static int open_keyboard(void);
+static int open_mouse(void);
+static int read_mouse(int *dx, int *dy, int *bp);
+static void close_framebuffer(void);
+static void close_keyboard(void);
+static void close_mouse(void);
+static void blit_blend(Drawable *src, const Rect *srect, Drawable *dst, const Rect *drect);
+static void blit_srccopy(Drawable *src, const Rect *srect, Drawable *dst, const Rect *drect);
+
+#if DEBUG
+#define unassert(a)   if (!(a)) unassert_handler(#a,__FILE__, __LINE__)
+
+/* exit graphics mode for error message */
+static void unassert_handler(char *msg, char *file, int line)
+{
+    host_gui_deinit();
+    printf("Assertion failed: %s at %s:%d\n", msg, file, line);
+    exit(255);
+}
+
+static void catch_signals(int signo)
+{
+    host_gui_deinit();
+    printf("SIGNAL %d\n", signo);
+    exit(1);
+}
+#else
+#define unassert(a)
+#endif
+
+void host_gui_begin_composite(void)
+{
+}
+
+void host_gui_end_composite(void)
+{
+}
+
+void host_gui_flush(const Rect *r)
+{
+    Rect cr;
+    cr.x = 0;
+    cr.y = 0;
+    cr.w = fb.width;
+    cr.h = fb.height;
+    unassert(r);
+    blit_srccopy(&bb, &cr, &fb, &cr);
+}
+
+void host_gui_resize(uint64_t w, uint64_t h)
+{
+}
+
+/* return rect adjusted to current clip rectangle */
+static Rect *get_clip_rect(const Rect *rect)
+{
+    static Rect r;
+
+    r = *rect;
+    if (r.w < 1 || r.h < 1) return NULL;
+    r.w += r.x;
+    r.h += r.y;
+    if (r.x >= clip.w || r.y >= clip.h) return NULL;
+    if (r.w <= clip.x || r.h <= clip.y) return NULL;
+    if (clip.x > r.x) r.x = clip.x;
+    if (clip.y > r.y) r.y = clip.y;
+    if (r.w > clip.w) r.w = clip.w;
+    if (r.h > clip.h) r.h = clip.h;
+    r.w -= r.x;
+    r.h -= r.y;
+    return &r;
+}
+
+/* set global clipping rectangle */
+void host_gui_set_clip(const Rect *rect)
+{
+    /* store as x, y, x1, y1 */
+    if (rect) {
+        clip = *rect;
+        clip.w += clip.x;
+        clip.h += clip.y;
+        if (clip.x < 0) clip.x = 0;
+        if (clip.y < 0) clip.y = 0;
+        if (clip.w >= fb.width) clip.w = fb.width - 1;
+        if (clip.h >= fb.height) clip.h = fb.height - 1;
+    } else {
+        clip.x = 0;
+        clip.w = fb.width - 1;
+        clip.y = 0;
+        clip.h = fb.height - 1;
+    }
+}
+
+/* set color for Drawables */
+void host_gui_set_color(uint8_t r, uint8_t g, uint8_t b, uint8_t a)
+{
+    //FIXME premul colors with alpha ?
+    draw_color = (a << 24) + (r << 16) + (g << 8)  + b;
+}
+
```

Comment:
Let me check this, I initially used `r = a * r` as the premul, thus the FIXME. The correct premul would be `r = (a * r) >> 255`, correct? 

Note this driver is using the less-exact alpha blend operation `c = (a * c) >> 8` throughout, rather than the more accurate `c = (a * (c + 1)) >> 8`. This will result in one less than full brightness for a=255, but I wanted to have all colors match, regardless of the otherwise potential difference between some of the blit cases. I plan on thinking through this a bit more to decide how to move all calculations to the latter better case. A similar problem exists in the sample raw driver with its mixed use of alpha blend math.

Reply:
Fixed.